key/trie/trie.go

```diff
@@ -0,0 +1,323 @@
+// Package trie provides an implementation of a XOR Trie
+package trie
+
+import (
+	"errors"
+
+	"github.com/plprobelab/go-kademlia/key"
+)
+
+var ErrMismatchedKeyLength = errors.New("key length does not match existing keys")
+
+// Trie is a trie for equal-length bit vectors, which stores values only in the leaves.
+// A node may optionally hold data of type T
+// Trie node invariants:
+// (1) Either both branches are nil, or both are non-nil.
+// (2) If branches are non-nil, key must be nil.
+// (3) If both branches are leaves, then they are both non-empty (have keys).
+type Trie[T any] struct {
+	branch [2]*Trie[T]
+	key    key.KadKey
+	data   T
+}
+
+func New[T any]() *Trie[T] {
+	return &Trie[T]{}
+}
+
+func (tr *Trie[T]) Key() key.KadKey {
```

Comment:
Not sure about the implications but `Trie` conforms to the `NodeID` interface. Perhaps rather export the `Key` and `Data` fields?

Reply:
I unexprted Key and Data when adding the mutable Add/Remove methods. My thinking is that we want to limit the chances for users to corrupt the trie by overwriting fields like `key`. In practice the Key method will be inlined so there's no real performance overhead.

Not sure I understand the point about conforming to NodeID interface - what would be the use case?